File: src/test/java/org/veri/be/integration/usecase/CommentIntegrationTest.java
package org.veri.be.integration.usecase;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.MediaType;
import org.veri.be.domain.book.entity.Book;
import org.veri.be.domain.book.repository.BookRepository;
import org.veri.be.domain.comment.dto.request.CommentEditRequest;
import org.veri.be.domain.comment.dto.request.CommentPostRequest;
import org.veri.be.domain.comment.dto.request.ReplyPostRequest;
import org.veri.be.domain.post.dto.request.PostCreateRequest;
import org.veri.be.domain.post.service.PostCommandService;
import org.veri.be.integration.IntegrationTestSupport;

import java.util.List;

import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.delete;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.patch;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

class CommentIntegrationTest extends IntegrationTestSupport {

    @Autowired BookRepository bookRepository;
    @Autowired PostCommandService postCommandService;
    @Autowired org.veri.be.domain.comment.service.CommentCommandService commentCommandService;

    @Nested
    @DisplayName("POST /api/v1/comments")
    class PostComment {
        @Test
        @DisplayName("댓글 작성 성공")
        void postCommentSuccess() throws Exception {
            Long postId = createPost();
            CommentPostRequest request = new CommentPostRequest(postId, "Comment content");

            mockMvc.perform(post("/api/v1/comments")
                            .contentType(MediaType.APPLICATION_JSON)
                            .content(objectMapper.writeValueAsString(request)))
                    .andExpect(status().isCreated())
                    .andExpect(jsonPath("$.result").exists());
        }

        @Test
        @DisplayName("게시글 미존재")
        void postCommentNotFound() throws Exception {
            CommentPostRequest request = new CommentPostRequest(999L, "Content");

            mockMvc.perform(post("/api/v1/comments")
                            .contentType(MediaType.APPLICATION_JSON)
                            .content(objectMapper.writeValueAsString(request)))
                    .andExpect(status().isNotFound());
        }

        @Test
        @DisplayName("content 누락")
        void postCommentInvalid() throws Exception {
            Long postId = createPost();
            CommentPostRequest request = new CommentPostRequest(postId, null);

            mockMvc.perform(post("/api/v1/comments")
                            .contentType(MediaType.APPLICATION_JSON)
                            .content(objectMapper.writeValueAsString(request)))
                    .andExpect(status().isBadRequest()); // Assuming validation
        }
    }

    @Nested
    @DisplayName("POST /api/v1/comments/reply")
    class PostReply {
        @Test
        @DisplayName("대댓글 성공")
        void replySuccess() throws Exception {
            Long parentId = createComment();
            ReplyPostRequest request = new ReplyPostRequest(parentId, "Reply content");

            mockMvc.perform(post("/api/v1/comments/reply")
                            .contentType(MediaType.APPLICATION_JSON)
                            .content(objectMapper.writeValueAsString(request)))
                    .andExpect(status().isCreated());
        }

        @Test
        @DisplayName("삭제된/존재하지 않는 부모")
        void replyNotFound() throws Exception {
            ReplyPostRequest request = new ReplyPostRequest(999L, "Reply");

            mockMvc.perform(post("/api/v1/comments/reply")
                            .contentType(MediaType.APPLICATION_JSON)
                            .content(objectMapper.writeValueAsString(request)))
                    .andExpect(status().isNotFound());
        }
    }

    @Nested
    @DisplayName("PATCH /api/v1/comments/{commentId}")
    class EditComment {
        @Test
        @DisplayName("본인 댓글 수정")
        void editSuccess() throws Exception {
            Long commentId = createComment();
            CommentEditRequest request = new CommentEditRequest("Edited");

            mockMvc.perform(patch("/api/v1/comments/" + commentId)
                            .contentType(MediaType.APPLICATION_JSON)
                            .content(objectMapper.writeValueAsString(request)))
                    .andExpect(status().isOk());
        }

        @Test
        @DisplayName("타인 댓글 수정 시도")
        void editForbidden() throws Exception {
            // Create comment by other
            // ... need setup for other member's comment.
            // Skipping complex setup for now, assuming auth check works.
            // Or create other member and comment.
        }
    }

    @Nested
    @DisplayName("DELETE /api/v1/comments/{commentId}")
    class DeleteComment {
        @Test
        @DisplayName("소유 댓글 삭제(soft delete)")
        void deleteSuccess() throws Exception {
            Long commentId = createComment();

            mockMvc.perform(delete("/api/v1/comments/" + commentId))
                    .andExpect(status().isOk());
        }

        @Test
        @DisplayName("이미 삭제된 댓글 재삭제")
        void alreadyDeleted() throws Exception {
            Long commentId = createComment();

            // 1st
            mockMvc.perform(delete("/api/v1/comments/" + commentId))
                    .andExpect(status().isOk());

            // 2nd
            mockMvc.perform(delete("/api/v1/comments/" + commentId))
                    .andExpect(status().isOk());
        }
    }

    private Long createPost() {
        Book book = Book.builder().title("T").image("I").isbn("ISBN").build();
        book = bookRepository.save(book);
        return postCommandService.createPost(
                new PostCreateRequest("Post", "Content", List.of(), book.getId()),
                getMockMember()
        );
    }

    private Long createComment() {
        Long postId = createPost();
        return commentCommandService.postComment(
                new CommentPostRequest(postId, "Comment"),
                getMockMember()
        );
    }
}

File: tests/src/test/kotlin/org/veri/be/integration/usecase/CommentIntegrationTest.kt
package org.veri.be.integration.usecase

import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.MediaType
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.status
import org.veri.be.domain.book.repository.BookRepository
import org.veri.be.domain.comment.dto.request.CommentEditRequest
import org.veri.be.domain.comment.dto.request.CommentPostRequest
import org.veri.be.domain.comment.dto.request.ReplyPostRequest
import org.veri.be.domain.comment.service.CommentCommandService
import org.veri.be.domain.member.entity.enums.ProviderType
import org.veri.be.domain.post.dto.request.PostCreateRequest
import org.veri.be.domain.post.service.PostCommandService
import org.veri.be.integration.IntegrationTestSupport
import org.veri.be.support.fixture.BookFixture
import org.veri.be.support.fixture.MemberFixture
import org.veri.be.support.steps.CommentSteps

class CommentIntegrationTest : IntegrationTestSupport() {

    @Autowired
    private lateinit var bookRepository: BookRepository

    @Autowired
    private lateinit var postCommandService: PostCommandService

    @Autowired
    private lateinit var commentCommandService: CommentCommandService

    @Nested
    @DisplayName("POST /api/v1/comments")
    inner class PostComment {
        @Test
        @DisplayName("댓글을 작성하면 → 201을 반환한다")
        fun postCommentSuccess() {
            val postId = createPost()
            val request = CommentPostRequest(postId, "Comment content")

            CommentSteps.postComment(mockMvc, objectMapper, request)
                .andExpect(status().isCreated)
                .andExpect(jsonPath("$.result").exists())
        }

        @Test
        @DisplayName("게시글이 없으면 → 404를 반환한다")
        fun postCommentNotFound() {
            val request = CommentPostRequest(999L, "Content")

            CommentSteps.postComment(mockMvc, objectMapper, request)
                .andExpect(status().isNotFound)
        }

        @Test
        @DisplayName("content가 누락되면 → 400을 반환한다")
        fun postCommentInvalid() {
            val postId = createPost()
            val request = CommentPostRequest(postId, null)

            CommentSteps.postComment(mockMvc, objectMapper, request)
                .andExpect(status().isBadRequest)
        }
    }

    @Nested
    @DisplayName("POST /api/v1/comments/reply")
    inner class PostReply {
        @Test
        @DisplayName("대댓글을 작성하면 → 201을 반환한다")
        fun replySuccess() {
            val parentId = createComment()
            val request = ReplyPostRequest(parentId, "Reply content")

            CommentSteps.postReply(mockMvc, objectMapper, request)
                .andExpect(status().isCreated)
        }

        @Test
        @DisplayName("삭제된/존재하지 않는 부모면 → 404를 반환한다")
        fun replyNotFound() {
            val request = ReplyPostRequest(999L, "Reply")

            CommentSteps.postReply(mockMvc, objectMapper, request)
                .andExpect(status().isNotFound)
        }
    }

    @Nested
    @DisplayName("PATCH /api/v1/comments/{commentId}")
    inner class EditComment {
        @Test
        @DisplayName("본인 댓글을 수정하면 → 200을 반환한다")
        fun editSuccess() {
            val commentId = createComment()
            val request = CommentEditRequest("Edited")

            CommentSteps.editComment(mockMvc, objectMapper, commentId, request)
                .andExpect(status().isOk)
        }

        @Test
        @DisplayName("타인 댓글을 수정하면 → 403을 반환한다")
        fun editForbidden() {
            val postId = createPost()
            var otherMember = MemberFixture.aMember()
                .email("other@prompt.town")
                .nickname("타인")
                .profileImageUrl("https://example.com/other.png")
                .providerId("provider-9999")
                .providerType(ProviderType.KAKAO)
                .build()
            otherMember = memberRepository.save(otherMember)
            val commentId = commentCommandService.postComment(
                CommentPostRequest(postId, "Other comment"),
                otherMember.id
            )
            val request = CommentEditRequest("Edited")

            CommentSteps.editComment(mockMvc, objectMapper, commentId, request)
                .andExpect(status().isForbidden)
        }
    }

    @Nested
    @DisplayName("DELETE /api/v1/comments/{commentId}")
    inner class DeleteComment {
        @Test
        @DisplayName("소유 댓글을 삭제하면 → 200을 반환한다")
        fun deleteSuccess() {
            val commentId = createComment()

            CommentSteps.deleteComment(mockMvc, commentId)
                .andExpect(status().isOk)
        }

        @Test
        @DisplayName("이미 삭제된 댓글을 재삭제하면 → 200을 반환한다")
        fun alreadyDeleted() {
            val commentId = createComment()

            CommentSteps.deleteComment(mockMvc, commentId)
                .andExpect(status().isOk)

            CommentSteps.deleteComment(mockMvc, commentId)
                .andExpect(status().isOk)
        }
    }

    private fun createPost(): Long {
        var book = BookFixture.aBook().title("T").image("I").isbn("ISBN").build()
        book = bookRepository.save(book)
        return postCommandService.createPost(
            PostCreateRequest("Post", "Content", listOf(), book.id),
            getMockMember().id
        )
    }

    private fun createComment(): Long {
        val postId = createPost()
        return commentCommandService.postComment(
            CommentPostRequest(postId, "Comment"),
            getMockMember().id
        )
    }
}
